### PR TITLE
UI, Icon: Render all Icons as img-tag

### PR DESCRIPTION
**Pull request by nhaagen**
This removes the surrounding div of icons and renders them, standard and custom, as img.
The aria-label is substituted by the alt-attribute according to https://github.com/ILIAS-eLearning/ILIAS/pull/2518.
The abbreviation-functions are still in the code, but do not show in the rendering besides a  data-attribute. In the future, the text might be amended to the default-svg.

### CODE REVIEW

File: src/UI/Component/Symbol/Icon/Standard.php [hunk 1]
```diff
@@ -127,17 +132,18 @@ interface Standard extends Icon
     const CMIS = 'cmis';	//xAPI/cmi5
     const REP = 'rep';	//Repository
     const TASK = 'task';   //Task
+    const LS = 'lso';   //Learning Sequence
```
Comment (klees) [inline src/UI/Component/Symbol/Icon/Standard.php]:
```suggestion
    const LSO = 'lso';   //Learning Sequence
```
Let's try to stick with "LSO" =)

File: src/UI/Component/Symbol/Icon/Standard.php [hunk 2]
```diff
@@ -8,6 +9,10 @@
  */
 interface Standard extends Icon
 {
+    const DEFAULT_ICON_NAME = 'default';
+    const ICON_PATH = './templates/default/images';
+    const ICON_OUTLINED_PATH = './templates/default/images/outlined';
+    const ICON_NAME_PATTERN = 'icon_%s.svg';
```
Comment (klees) [inline src/UI/Component/Symbol/Icon/Standard.php]: These should go into the `Renderer` imo, because they clearly are a detail of the rendering and don't belong to the general UI component.

File: src/UI/Component/Symbol/Icon/Standard.php [hunk 3]
```diff
@@ -127,17 +132,18 @@ interface Standard extends Icon
     const CMIS = 'cmis';	//xAPI/cmi5
     const REP = 'rep';	//Repository
     const TASK = 'task';   //Task
+    const LS = 'lso';   //Learning Sequence
 
     /**
      * Is this an outlined Icon?
      */
-    public function isOutlined();
+    public function isOutlined() : bool;
```
Comment (klees) [inline src/UI/Component/Symbol/Icon/Standard.php:135]: :+1:

File: src/UI/Component/Symbol/Icon/Standard.php [hunk 4]
```diff
@@ -127,17 +132,18 @@ interface Standard extends Icon
...
 
     /**
      * Get an icon like this, but marked as outlined.
      *
      * @param bool $is_outlined
      * @return Standard
```
Comment (klees) [inline src/UI/Component/Symbol/Icon/Standard.php]:
```suggestion
```

File: src/UI/Implementation/Component/Symbol/Icon/Custom.php [hunk 5]
```diff
@@ -15,10 +15,8 @@ class Custom extends Icon implements C\Symbol\Icon\Custom
     private $icon_path;
 
 
-    public function __construct($icon_path, $aria_label, $size, $is_disabled)
+    public function __construct(string $icon_path, string $label, string $size, bool $is_disabled)
     {
-        $this->checkStringArg("string", $icon_path);
-        $this->checkStringArg("string", $aria_label);
```
Comment (klees) [inline src/UI/Implementation/Component/Symbol/Icon/Custom.php:21]: :+1:

File: src/UI/Implementation/Component/Symbol/Icon/Custom.php [hunk 6]
```diff
@@ -28,7 +26,7 @@ public function __construct($icon_path, $aria_label, $size, $is_disabled)
         $this->checkBoolArg("is_disabled", $is_disabled);
```
Comment (klees) [inline src/UI/Implementation/Component/Symbol/Icon/Custom.php]:
```suggestion
```

File: src/GlobalScreen/Scope/MainMenu/Factory/hasSymbolTrait.php [hunk 7]
```diff
@@ -19,8 +19,8 @@ trait hasSymbolTrait
     public function withSymbol(Symbol $symbol) : hasSymbol
     {
         // bugfix mantis 25526: make aria labels mandatory
-        if (($symbol instanceof Icon\Icon || $symbol instanceof Glyph\Glyph)
-            && ($symbol->getAriaLabel() === "")) {
+        if (($symbol instanceof Glyph\Glyph && $symbol->getAriaLabel() === "") ||
+            ($symbol instanceof Icon\Icon && $symbol->getLabel() === "")) {
             throw new \LogicException("the symbol's aria label MUST be set to ensure accessibility");
         }
```
Comment (klees) [inline src/GlobalScreen/Scope/MainMenu/Factory/hasSymbolTrait.php:25]: Please be so kind and add an according entry in the roadmap of the UI-framework. I think that this constraint should be enforced by the UI-framework.

File: src/UI/Implementation/Component/Symbol/Icon/Renderer.php [hunk 8]
```diff
@@ -38,17 +38,9 @@ public function render(Component\Component $component, RendererInterface $defaul
             $tpl->setVariable("ABBREVIATION", $ab);
         }
 
-        $di = $component->isDisabled();
-        if ($di) {
-            $tpl->setVariable("DISABLED", " disabled");
-        }
-
-        $id = $this->bindJavaScript($component);
```
Comment (klees) [inline src/UI/Implementation/Component/Symbol/Icon/Renderer.php:46]: Why did you remove this? I think we still need the ability to bind JS...

File: src/UI/Implementation/Component/Symbol/Icon/Standard.php [hunk 9]
```diff
@@ -169,10 +167,24 @@ public function isOutlined() : bool
...
+        if (!in_array($name, $this->getAllStandardHandles())) {
+            $name = self::DEFAULT_ICON_NAME;
+        }
+
+        $path = self::ICON_PATH;
+        if ($this->isOutlined()) {
+            $path = self::ICON_OUTLINED_PATH;
+        }
+        return $path . '/' . sprintf(self::ICON_NAME_PATTERN, $name);
+    }
```
Comment (klees) [inline src/UI/Implementation/Component/Symbol/Icon/Standard.php]: As with the according constants, I think this is a concern of the renderer not the component itself.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @nhaagen,

LGTM in general and especially nice that we have lost so much LOC. Besides my inline change requests: Did you try this changes in some installation?

Best regards!

File: src/UI/Implementation/Component/Symbol/Icon/Renderer.php (hunk 8 above)
Comment (nhaagen) [reply src/UI/Implementation/Component/Symbol/Icon/Renderer.php:46]: When (re)reading the Icons' description, I found "Icons themselves are not interactive; however they are allowed  within interactive containers."
What kind of JS would you fathom?

File: src/UI/Implementation/Component/Symbol/Icon/Standard.php (hunk 9 above)
Comment (nhaagen) [reply src/UI/Implementation/Component/Symbol/Icon/Standard.php]: you are so right, thanks ;)

**Comment by nhaagen** [comment]
"Did you try this changes in some installation?" - Other than my local one? No.

**Comment by klees** [comment]
@chfsx @alex40724 I think we had some discussion in mantis regarding the binding of javascript to icons of slates, but I can't find it. If you (still) need the possiblity to bind javascript to icons, please step in here!

**Review by klees — APPROVED** [review]
@Amstutz Please have a look as well and maybe contact me to discuss how we want to proceed here. Note that we would need to make this change in release_6 as well to improve accessibility.

**Comment by chfsx** [comment]
Hi @klees yes the possibility to bind js to all type of icons is still needed, are all Icons JavaScriptBindable with this PR then?

**Comment by nhaagen** [comment]
Hi @chfsx , with this PR I actually removed JSBinding from Icons. If i's still needed, I will re-add, of course. Could you point me to an example where there is JS on an Icon? Probably the JS has to be adjusted as well...

**Comment by Amstutz** [comment]
Hi @nhaagen 

I am fine with the proposed changes in the interface and the description. Thx for cleaning this up, now it reflects the current situation much better. 

Question:
- [x] Performance: While reviewing this, I observed that unfortunately, this will be quite a bit slower than the old solution, since the image path was giving statically by CSS. Now we will have to resolve the path for every icon dynamically (see also point bellow). Do you agree that this will have some negative impact? 

Should Change, or give some reasoning why not:
- [ ] Binding: If we find valid cases for Icons using the binding, we should keep it. This PR is not mainly about removing those, so this might be for another day. However, we should at least have one such case.

Please Change:
- [ ] Template Variable: You still use "Customimage" for both, the custom icon and the standard ones. Please change to just Image or something along that line.
- [ ] Path Resolver: This is very tricky. The current implementation of getStandardIconPath will not do. Currently one would have not option anymore to customize the icon set without changing the core. This worked with the previous system, since we had something like image path in less which could easily be changed by the skin. Currently the static getImagePath of our beloved ilUtil is used for this in other cases. Note that this static function has deps to ilStyleDefinition (also used statically inside) which again has deps to the DB (the used style and image folder that comes with it can only be resolved by accessing the DB due to locally applied sub styles). So this is a rather big bowl of stinky mud nobody had time to refactor yet (see also comment of shame in the beginning of ./Services/Style/System/classes/class.ilStyleDefinition.php). So how do we deal with that here? Probably cleanest way is to simply reverse the dependency here. Meaning, propose an interface, like "ImagePathResolver" or similar which has probably exactly one method "resolveImagePath" and inject it in the initialization of the UI Framework with some local class calling ilUtil. Obviously this is still a bit a mess since we would really need to move that logic from ilUtil to SystemStyles and remove all the static mess out of ilStyleDefinition. However, this is outside of your area of concern.

Thanks a lot. 

@klees, since this "ImagePathResolver" will need to be on the public, plz shortly give some feedback on this suggestion. If you are fine with it, you can add the JF label, so we can discuss the proposed public changes in the upcoming jf.

Thanks to all involved
@Amstutz

**Comment by klees** [comment]
Hi @Amstutz,

I totally agree regarding the `ImageResolver` and just didn't noticed that missing customizabiliy yet. I'm glad you spotted this, thanks. Reversing that dependency as you suggest seems to be the way forward, regarding that problems in the general implementation of that mechanism. I think you are also correct regarding the performance, but I'm not too much into the actual mechanism of resolving to guess how much this will really cost. Having an `ImageResolver` would be of help here too, since we could easily use it to mitigate that issue by implementing a version using caching if performance turns out to be a problem.

Best regards!

**Comment by nhaagen** [comment]
Hi @Amstutz,
good catch spotting that PathResolver-thing. Thank you.
Did you have something like this in mind?

**Review by Amstutz — APPROVED** [review]
Yes, thx a lot. This does the trick. @klees feel free to merge as this is good from your POV.

**Comment by klees** [comment]
Hi @Amstutz,

currently I'm insecure here. The a11y report seems to contain more requirements for the icons then we knew when starting to implement this, so I guess we should also wait till thursday.

Best regards!

**Comment by Amstutz** [comment]
Any insight on this from thursday?

**Comment by Amstutz** [comment]
ping

**Comment by Amstutz** [comment]
how do we proceed with this?

**Comment by klees** [comment]
Hi @Amstutz,

I'll ask the SIG Accessibility for opinion and funding and hope to have an answer after my vacation.

If not, I guess we should just go on with this as is, seems to be some win anyway.

Best regards!

File: src/UI/Component/Symbol/Factory.php [hunk 10]
```diff
@@ -46,11 +42,10 @@ interface Factory
      *     1: Icons MUST have a class indicating their usage.
      *     2: Icons MUST be tagged with a CSS-class indicating their size.
      *   accessibility:
-     *     1: Icons MUST bear an aria-label.
+     *     1: Icons MUST bear an alt-text.
      *     2: Disabled Icons MUST bear an aria-label indicating the disabled status.
      *   wording:
-     *     1: The aria-label MUST state the represented object-type.
-     *     2: The abbreviation SHOULD consist of one or two letters.
+     *     1: The alt-text MUST state the represented object-type.
```
Comment (atoedt) [inline src/UI/Component/Symbol/Factory.php:48]: i.e. File or Course.

**Comment by Amstutz** [comment]
Any update here?

**Comment by klees** [comment]
Hi @Amstutz,

I got an answer from @atoedt in the meantime, I'll try to look into this asap.

Best regards!

**Comment by klees** [comment]
@Amstutz and I talked about this today, this is still on my agenda. Since this is currently volunteer work and not funded by anyone, it just has no specific priority for me. Since we have a ticket here, we could well tackle this during beta. If anyone wants to provide funding for this accessibility requirement, feel free to contact me.

**Comment by klees** [comment]
Thanks to Uni Bern for funding further steps here. This is superseeded by  #3143.
